### PR TITLE
Polishing and Adding to Options Menu

### PR DESCRIPTION
**Pull request by BrickPi**
Fixed up a lot of the UI for the options menu, added some missing reqs. This is only a beginning, once we have actual UI elements for the options menu it needs to be added in, and the settings are currently visual-only, we need to look at designing a singleton settings state object from which to pull the users preferences in both gdscript and C++.

This PR covers UI-48, UI-50, UI-15, UI-18, UI-21, UIFUN-19, UIFUN-20, UIFUN-22, & UIFUN-45

### CODE REVIEW

**Review by Orwellian-225 — COMMENTED** [review]
Looks great

**Review by Orwellian-225 — APPROVED** [review]

File: game/src/OptionMenu/VideoTab.gd [hunk 1]
```diff
@@ -0,0 +1,6 @@
+extends HBoxContainer
+
+@export var initial_focus: Button
+
+func _ready():
+	initial_focus.grab_focus()
```
Comment (Spartan322) [inline game/src/OptionMenu/VideoTab.gd:6]: This accomplishes nothing because `_ready` does not run on visibility, it runs on being loaded into the scene tree but since this can't be seen nothing can be focused.

**Review by Spartan322 — CHANGES_REQUESTED** [review]
Avoid using non-container nodes for container purposes, Panel should be replaced with PanelContainer and should contain the OptionsMenu tab

Do not put themes on child scenes, always keep them at the highest scene level, until godotengine/godot#74267 is fixed assign it in the OptionsMenu scene. (When godotengine/godot#74267 is fixed in a following Godot release we'll just be assigning the theme to the ProjectSettings)

LocaleButton's text is not aligned with the other selector buttons.

GeneralTab does not follow the patterns of the other tabs:

- Anchors are not used as they don't serve a purpose here
- There is no child VBoxContainer for an vertical size expanded Control spacer and GridContainer node
- There is no top Control spacer node with a `size_flags_stretch_ratio` of 0.1 (found under Container Sizes Stretch Ratio)

File: game/src/GameMenu.tscn [hunk 2]
```diff
@@ -1,10 +1,9 @@
-[gd_scene load_steps=7 format=3 uid="uid://b4pg2y2ivib8f"]
+[gd_scene load_steps=6 format=3 uid="uid://vbtt55n875k1"]
 
 [ext_resource type="Script" path="res://src/GameMenu.gd" id="1_cafwe"]
 [ext_resource type="PackedScene" uid="uid://dvoin538iby54" path="res://src/MainMenu/MainMenu.tscn" id="2_2jbkh"]
 [ext_resource type="PackedScene" uid="uid://cnbfxjy1m6wja" path="res://src/OptionMenu/OptionsMenu.tscn" id="3_111lv"]
-[ext_resource type="PackedScene" uid="uid://b7oncobnacxmt" path="res://src/LocaleButton.tscn" id="4_jno35"]
```
[the remark this answers is not in the document]
Comment (BrickPi) [reply game/src/GameMenu.tscn:6]: I mean, it makes no difference, if they don't know any of the language they're not going to be able to find the locale button on the main menu rather than the options menu, and assuming we ship the manual in multiple languages, one of the first sections can simply be where the options menu is and how to change the language.

File: game/src/OptionMenu/GeneralTab.gd [hunk 3]
```diff
@@ -0,0 +1,6 @@
+extends HBoxContainer
+
+@export var initial_focus: Button
+
+func _ready():
+	initial_focus.grab_focus()
```
[the remark this answers is not in the document]
Comment (BrickPi) [reply game/src/OptionMenu/GeneralTab.gd:6]: @Spartan322 is there a method that is called on visibility? For accessibility it feels good to grab focus of the first button and assign tabstops.

[the remark this answers is not in the document]
Comment (Spartan322) [reply game/src/OptionMenu/GeneralTab.gd:6]: There's two ways, [`CanvasItem.NOTIFICATION_VISIBILITY_CHANGED`](https://docs.godotengine.org/en/stable/classes/class_canvasitem.html#constants) and [`CanvasItem.visibility_changed()` signal](https://docs.godotengine.org/en/stable/classes/class_canvasitem.html#signals), inside a node of itself I would recommend sticking to the notification, how we use it is like follows:
```gdscript
func _notification(what : int) -> void:
	match(what):
		NOTIFICATION_VISIBILITY_CHANGED:
			if visible: initial_focus.grab_focus()
```

File: game/src/GameMenu.tscn (hunk 2 above)
[the remark this answers is not in the document]
Comment (Spartan322) [reply game/src/GameMenu.tscn:6]: That's requiring excessive steps and other people involved, which is dumb, nobody should require outside help just to start the game, a locale button at start will eliminate this if we put an icon for each language name.

File: game/src/OptionMenu/GeneralTab.tscn [hunk 4]
```diff
@@ -6,30 +6,29 @@
...
+[node name="VBoxContainer" type="VBoxContainer" parent="."]
+layout_mode = 2
+
+[node name="Control" type="Control" parent="VBoxContainer"]
+layout_mode = 2
+size_flags_vertical = 3
+size_flags_stretch_ratio = 0.1
+
+[node name="GridContainer" type="GridContainer" parent="VBoxContainer"]
 layout_mode = 2
```
Comment (Spartan322) [inline game/src/OptionMenu/GeneralTab.tscn:22]:
```suggestion
layout_mode = 2
size_flags_vertical = 3
```

**Review by Spartan322 — CHANGES_REQUESTED** [review]

**Comment by Spartan322** [comment]
Once that one detail is fixed, a rebase conflict resolve should be done.

**Review by Spartan322 — CHANGES_REQUESTED** [review]
Now just needs a squash rebase to one commit and to resolve the conflicts.

**Comment by BrickPi** [comment]
That should be it, unless I need to re-rebase it due to the merge artefact files deleted in b96fde2.

**Review by Spartan322 — CHANGES_REQUESTED** [review]
This is not how you're supposed to do a merge or a rebase, and I did say this should've been a rebase that was squashed down to a single commit, these commits cannot be merged like this, it causes massive problems to the branch.

**Comment by BrickPi** [comment]
Should be all ready to merge.

File: game/src/MainMenu/MainMenu.gd [hunk 5]
```diff
@@ -16,12 +16,14 @@ var _checksum_label : Label
...
-	_new_game_button.grab_focus()
-
+	# UI-111
+	_checksum_label.tooltip_text = "Checksum " + Checksum.get_checksum_text()
+	_checksum_label.text = "(" + Checksum.get_checksum_text().substr(0, 4) + ")"
+	
+func _notification(what : int) -> void:
+	match(what):
+		NOTIFICATION_VISIBILITY_CHANGED:
+			if visible: _new_game_button.grab_focus()
```
Comment (Spartan322) [inline game/src/MainMenu/MainMenu.gd:26]:
```suggestion
	_on_new_game_button_visibility_changed()
```

And at the bottom you have:
```gdscript
func _on_new_game_button_visibility_changed():
	if _new_game_button.visible:
		_new_game_button.grab_focus.call_deferred()
```

File: game/src/MainMenu/MainMenu.tscn [hunk 6]
```diff
@@ -152,6 +153,12 @@ mouse_filter = 1
 theme_type_variation = &"Label_Checksum"
 text = "(0000)"
 
+[node name="LocaleButton" parent="Panel/VBox/Margin2" instance=ExtResource("3_amonp")]
+layout_mode = 2
+size_flags_horizontal = 8
+alignment = 0
+text_overrun_behavior = 4
+
 [connection signal="pressed" from="Panel/VBox/Margin/ButtonList/NewGameButton" to="." method="_on_new_game_button_pressed"]
```
Comment (Spartan322) [inline game/src/MainMenu/MainMenu.tscn:162]:
```suggestion
[connection signal="pressed" from="Panel/VBox/Margin/ButtonList/NewGameButton" to="." method="_on_new_game_button_pressed"]
[connection signal="visibility_changed" from="Panel/VBox/Margin/ButtonList/NewGameButton" to="." method="_on_new_game_button_visibility_changed"]
```

**Review by Spartan322 — CHANGES_REQUESTED** [review]

File: game/src/OptionMenu/OptionsMenu.tscn [hunk 7]
```diff
@@ -1,43 +1,41 @@
...
-theme_override_constants/margin_right = 250
-theme_override_constants/margin_bottom = 200
+layout_mode = 2
+theme_override_constants/margin_left = 180
+theme_override_constants/margin_top = 150
+theme_override_constants/margin_right = 180
+theme_override_constants/margin_bottom = 150
 
 [node name="Tab" type="TabContainer" parent="Margin"]
-editor_description = "UI-45"
```
Comment (Spartan322) [inline game/src/OptionMenu/OptionsMenu.tscn:33]: Requirements description being removed.

File: game/src/OptionMenu/OptionsMenu.tscn [hunk 8]
```diff
@@ -1,43 +1,41 @@
...
 layout_mode = 2
-size_flags_vertical = 3
```
Comment (Spartan322) [inline game/src/OptionMenu/OptionsMenu.tscn:35]: Is this superfluous or unnecessary?

File: game/src/OptionMenu/OptionsMenu.tscn [hunk 9]
```diff
@@ -1,43 +1,41 @@
...
 tab_alignment = 1
 use_hidden_tabs_for_min_size = true
 
+[node name="General" parent="Margin/Tab" instance=ExtResource("3_6gvf6")]
+editor_description = "UI-48"
+layout_mode = 2
+
 [node name="Video" parent="Margin/Tab" instance=ExtResource("2_ji8xr")]
-editor_description = "UI-46, UIFUN-43"
```
Comment (Spartan322) [inline game/src/OptionMenu/OptionsMenu.tscn:40]: Requirements description being removed.

File: game/src/OptionMenu/OptionsMenu.tscn [hunk 10]
```diff
@@ -1,43 +1,41 @@
...
-[ext_resource type="PackedScene" uid="uid://b7oncobnacxmt" path="res://src/LocaleButton.tscn" id="2_d7wvq"]
+[ext_resource type="Theme" uid="uid://cr4lh0vraucx7" path="res://default_theme.tres" id="2_8cfng"]
 [ext_resource type="PackedScene" uid="uid://bq3awxxjn1tuw" path="res://src/OptionMenu/VideoTab.tscn" id="2_ji8xr"]
 [ext_resource type="PackedScene" uid="uid://cbtgwpx2wxi33" path="res://src/OptionMenu/SoundTab.tscn" id="3_4w35t"]
+[ext_resource type="PackedScene" uid="uid://duwjal7sd7p6w" path="res://src/OptionMenu/GeneralTab.tscn" id="3_6gvf6"]
 [ext_resource type="PackedScene" uid="uid://bq7ibhm0txl5p" path="res://addons/keychain/ShortcutEdit.tscn" id="4_vdhjp"]
 [ext_resource type="PackedScene" uid="uid://dp2grvybtecqu" path="res://src/OptionMenu/OtherTab.tscn" id="5_ahefp"]
 
-[node name="OptionsMenu" type="Control"]
-editor_description = "UI-25"
```
Comment (Spartan322) [inline game/src/OptionMenu/OptionsMenu.tscn:11]: Requirements description being removed.

**Review by Spartan322 — CHANGES_REQUESTED** [review]
Looks almost good, anywhere else that's lacking requirement description?

File: game/src/OptionMenu/OptionsMenu.tscn (hunk 8 above)
Comment (BrickPi) [reply game/src/OptionMenu/OptionsMenu.tscn:35]: Pretty sure that was added by editor, I'm not going to touch it cause it won't do any harm to keep.

File: game/src/GameMenu.tscn [hunk 11]
```diff
@@ -43,25 +42,6 @@ layout_mode = 1
...
-grow_horizontal = 0
-grow_vertical = 0
-alignment = 2
-
-[node name="LocaleButton" parent="HBox" instance=ExtResource("4_jno35")]
-layout_mode = 2
-
-[connection signal="credits_button_pressed" from="MainMenu" to="." method="_on_main_menu_credits_button_pressed"]
-[connection signal="new_game_button_pressed" from="MainMenu" to="." method="_on_main_menu_new_game_button_pressed"]
-[connection signal="options_button_pressed" from="MainMenu" to="." method="_on_main_menu_options_button_pressed"]
```
Comment (Spartan322) [inline game/src/GameMenu.tscn:64]: Why were these removed?

Comment (BrickPi) [reply game/src/GameMenu.tscn]: No clue, should be reverted now.

File: game/src/OptionMenu/OptionsMenu.tscn [hunk 12]
```diff
@@ -49,25 +49,7 @@ layout_mode = 2
...
-mouse_filter = 2
-alignment = 2
-
-[node name="LocaleHBox" type="HBoxContainer" parent="LocaleVBox"]
-layout_mode = 2
-mouse_filter = 2
-alignment = 2
-
-[node name="LocaleButton" parent="LocaleVBox/LocaleHBox" instance=ExtResource("2_d7wvq")]
-editor_description = "UI-79"
```
Comment (Spartan322) [inline game/src/OptionMenu/OptionsMenu.tscn:72]: This might need to be added back somewhere.

**Review by Spartan322 — DISMISSED** [review]
Looks good.

**Review by ClarkeCode — DISMISSED** [review]

File: game/src/OptionMenu/OptionsMenu.tscn (hunk 7 above)
Comment (Spartan322) [reply game/src/OptionMenu/OptionsMenu.tscn:35]: Shouldn't UI-45 still be here?

File: game/src/OptionMenu/OptionsMenu.gd [hunk 13]
```diff
@@ -40,11 +40,11 @@ func _notification(what):
 	match what:
 		NOTIFICATION_CRASH:
 			_on_window_close_requested()
-
-# Could pass the LocaleButton between the MainMenu and OptionsMenu
-# but that seems a bit excessive
-func toggle_locale_button_visibility(locale_visible : bool):
-	$LocaleVBox/LocaleHBox/LocaleButton.visible = locale_visible
+
```
Comment (Spartan322) [inline game/src/OptionMenu/OptionsMenu.gd:43]: Superfluous whitespace, should only be a new line:
```suggestion

```

File: game/src/OptionMenu/GeneralTab.tscn [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+script = ExtResource("2_t06tb")
+section_name = "General"
+setting_name = "Autosave Interval"
+default_selected = 0
+
+[node name="LocaleLabel" type="Label" parent="VBoxContainer/GridContainer"]
+layout_mode = 2
+text = "Language"
+
+[node name="LocaleButton" parent="VBoxContainer/GridContainer" instance=ExtResource("2_5cfd7")]
```
Comment (Spartan322) [inline game/src/OptionMenu/GeneralTab.tscn:76]:
```suggestion
[node name="LocaleButton" parent="VBoxContainer/GridContainer" instance=ExtResource("2_5cfd7")]
editor_description = "UI-79"
```

File: game/src/OptionMenu/GeneralTab.tscn [hunk 15]
```diff
@@ -0,0 +1,80 @@
+[gd_scene load_steps=5 format=3 uid="uid://duwjal7sd7p6w"]
+
+[ext_resource type="Script" path="res://src/OptionMenu/GeneralTab.gd" id="1_gbutn"]
+[ext_resource type="PackedScene" uid="uid://b7oncobnacxmt" path="res://src/LocaleButton.tscn" id="2_5cfd7"]
+[ext_resource type="Script" path="res://src/OptionMenu/SettingNodes/SettingOptionButton.gd" id="2_msx2u"]
+[ext_resource type="Script" path="res://src/OptionMenu/AutosaveIntervalSelector.gd" id="2_t06tb"]
+
+[node name="GeneralTab" type="HBoxContainer" node_paths=PackedStringArray("initial_focus")]
```
Comment (Spartan322) [inline game/src/OptionMenu/GeneralTab.tscn:8]: Should probably have:
```suggestion
[node name="GeneralTab" type="HBoxContainer" node_paths=PackedStringArray("initial_focus")]
editor_description = "UI-48"
```

File: game/src/OptionMenu/VideoTab.tscn [hunk 16]
```diff
@@ -1,12 +1,17 @@
...
 
 [ext_resource type="Script" path="res://src/OptionMenu/ResolutionSelector.gd" id="1_i8nro"]
+[ext_resource type="Script" path="res://src/OptionMenu/VideoTab.gd" id="1_jvv62"]
 [ext_resource type="Script" path="res://src/OptionMenu/ScreenModeSelector.gd" id="2_wa7vw"]
 [ext_resource type="Script" path="res://src/OptionMenu/MonitorDisplaySelector.gd" id="3_y6lyb"]
+[ext_resource type="Script" path="res://src/OptionMenu/RefreshRateSelector.gd" id="4_381mg"]
+[ext_resource type="Script" path="res://src/OptionMenu/QualityPresetSelector.gd" id="5_srg4v"]
 
-[node name="Video" type="HBoxContainer"]
+[node name="Video" type="HBoxContainer" node_paths=PackedStringArray("initial_focus")]
```
Comment (Spartan322) [inline game/src/OptionMenu/VideoTab.tscn:10]: Since we're already modifying it, and to fit previous suggestion:
```suggestion
[node name="Video" type="HBoxContainer" node_paths=PackedStringArray("initial_focus")]
editor_description = "UI-46"
```

**Review by Spartan322 — APPROVED** [review]
Few little details need to be corrected, but otherwise looks good, functionally seems fine.

**Review by Spartan322 — CHANGES_REQUESTED** [review]
The locale on the main menu and general tab still aren't synced.

File: game/src/LocaleButton.gd [hunk 17]
```diff
@@ -32,12 +37,17 @@ func load_setting(file : ConfigFile) -> void:
 	var load_value = file.get_value(section_name, setting_name, TranslationServer.get_locale())
 	match typeof(load_value):
 		TYPE_STRING, TYPE_STRING_NAME:
-			var locale_index := _locales_list.find(load_value as String)
-			if locale_index != -1:
-				selected = locale_index
+			if _select_locale_by_string(load_value as String):
```
Comment (Spartan322) [inline game/src/LocaleButton.gd:40]: This is necessary to load the locale from the setting:
```suggestion
			if _select_locale_by_string(load_value as String):
				item_selected.emit(selected)
```

**Review by Spartan322 — CHANGES_REQUESTED** [review]

**Review by Spartan322 — APPROVED** [review]
Everything should be good now, just gotta wait for the CI to merge.
